Add `VersionAssemblyName` to package manifest

ronaldbarendse opened the pull request:
### Prerequisites

- [x] I have added steps to test this contribution in the description below

### Description
Currently the version in the `package.manifest` needs to be manually set (and kept up to date!) if you want to have it shown on the installed packages overview in the back-office and included in the telemetry data.

In most cases the `package.manifest` is shipped alongside an assembly (DLL), which is already versioned and with the changes in this PR, you can leave the `version` empty (or remove it from the JSON) and set `versionAssemblyName` to your package assembly name to have it automatically populated with the assembly version.

This ensures more packages will report the (correct) version and makes it easier for package authors (one less step/thing not to forget). The default `dotnet new umbracopackage` template will also automatically set this new property. I've also fixed an issue where the installed package overview didn't show the package version if a package migration state is already present in the database.

------------------

To test: build/pack Umbraco from the repository root, install the newly created package template, create/pack a new Umbraco package project and add that to the Umbraco.Web.UI project using:
```
dotnet pack --configuration Release --output build.out
dotnet new install build.out\Umbraco.Templates.*.nupkg
dotnet new umbracopackage --name Umbraco.Community.UmbracoPackage --output build.out\Umbraco.Community.UmbracoPackage
dotnet pack build.out\Umbraco.Community.UmbracoPackage --configuration Release --output build.out -p:EnablePackageValidation=false
dotnet add src\Umbraco.Web.UI package Umbraco.Community.UmbracoPackage --source build.out --prerelease
````

Next up, build the Umbraco.Web.UI project and start the application. You'll see the version is populated (instead of Unknown):

![Packages - Installed](https://user-images.githubusercontent.com/1051287/229294645-dfdb144b-63be-43d7-adfe-42c303f60c47.png)

------------------

After testing, you can clean up using:
```
dotnet clean src\Umbraco.Web.UI
dotnet remove src\Umbraco.Web.UI package Umbraco.Community.UmbracoPackage
dotnet new uninstall Umbraco.Templates

# Optionally, if you also want the NuGet packages removed from your local cache:
dotnet nuget locals --clear all
```

ronaldbarendse commented on src/Umbraco.Infrastructure/Manifest/ManifestParser.cs:
```diff
@@ -217,6 +224,31 @@ public PackageManifest ParseManifest(string text)
 
+    private string? GetAssemblyVersion(string name)
+    {
+        foreach (Assembly assembly in AssemblyLoadContext.All.SelectMany(x => x.Assemblies))
```
I initially used `ITypeFinder.AssembliesToScan` (to have a cached list of assemblies), but that only includes assemblies that have references to Umbraco (since that's what `IAssemblyProvider.Assemblies` returns in the `DefaultUmbracoAssemblyProvider` implementation). This contains the following comment:

https://github.com/umbraco/Umbraco-CMS/blob/60a5b19dc9fe24f5379a253e1f3fdcb9cb537dfc/src/Umbraco.Core/Composing/DefaultUmbracoAssemblyProvider.cs#L32-L37

So to have access to all assemblies, I've used `AssemblyLoadContext.All` for now, but since package assemblies are loaded on start-up (and not dynamically generated/loaded), it might be enough to change this to:

```suggestion
        foreach (Assembly assembly in AssemblyLoadContext.Default.Assemblies)
```

ronaldbarendse replied:
I can confirm `AssemblyLoadContext.Default` will have all the package assemblies 👍🏻

ronaldbarendse commented on src/Umbraco.Infrastructure/Services/Implement/PackagingService.cs:
```diff
@@ -157,14 +159,15 @@ public IEnumerable<InstalledPackage> GetAllInstalledPackages()
                 installedPackages.Add(package.PackageName, installedPackage);
             }
 
+            installedPackage.Version = string.IsNullOrEmpty(package.Version) ? "Unknown" : package.Version;
```
The version was previously only set when it couldn't find the package in the list, so it would be empty when it was already added when populating the list from the package migration states in the database.

ronaldbarendse replied:
And this is now also conditionally set, so the version is only shown when available (you won't see `Version: Unknown` in the dashboard anymore) 🎉

ronaldbarendse commented:
I've ensured the listed installed packages that only have a package migration plan will now use the version from the assembly they're defined in. So adding the following class into the example project (created above):

```c#
using Umbraco.Cms.Core.Packaging;

public class UmbracoPackage2MigrationPlan : PackageMigrationPlan
{
    public UmbracoPackage2MigrationPlan()
        : base("Umbraco.Community.UmbracoPackage2")
    {
    }

    protected override void DefinePlan()
    {
    }
}
```

Will result in listing `Umbraco.Community.UmbracoPackage2` with the same version:

![Installed packages](https://github.com/umbraco/Umbraco-CMS/assets/1051287/b0facaac-3582-4207-9436-8f9c3ab196cb)

bergmania commented on templates/UmbracoPackage/App_Plugins/UmbracoPackage/package.manifest:
```diff
@@ -1,5 +1,5 @@
-﻿{
+{
     "name": "UmbracoPackage",
-    "version": "",
```
Guess we need this is the new RCL template too

bergmania reviewed (approved):
Looks good, but we need to add the default to the RCL template too

ronaldbarendse commented:
> Looks good, but we need to add the default to the RCL template too

I've merged the changes that added the new Umbraco Package RCL template, updated that package manifest as well and updated the base branch to target the v12 release 👍🏻

bergmania commented:
Great @ronaldbarendse
